### PR TITLE
GST-9 - Initial commit, created quarkus-testcontainers submodule

### PR DESCRIPTION
**Pull request by pauldpearson**

### CODE REVIEW

File: quarkus-modules/quarkus-testcontainers/customer-service/src/test/java/info/customer/TestcontainersManager.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import io.quarkus.test.common.QuarkusTestResourceLifecycleManager;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.Network;
+import org.testcontainers.containers.PostgreSQLContainer;
+import org.testcontainers.containers.wait.strategy.Wait;
+import org.testcontainers.utility.DockerImageName;
+
+import java.util.Map;
+
+public class TestcontainersManager implements QuarkusTestResourceLifecycleManager {
```
Comment (theangrydev) [inline quarkus-modules/quarkus-testcontainers/customer-service/src/test/java/info/customer/TestcontainersManager.java:12]: Can we rename this to make it clear that it's a custom class not a default one

File: quarkus-modules/quarkus-testcontainers/customer-service/src/test/java/info/customer/CustomerResourceTest.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.stream.Stream;
+
+@QuarkusTestResource(TestcontainersManager.class)
+@QuarkusTest
+@TestHTTPEndpoint(CustomerResource.class)
+class CustomerResourceTest {
```
Comment (theangrydev) [inline quarkus-modules/quarkus-testcontainers/customer-service/src/test/java/info/customer/CustomerResourceTest.java:17]: We have a convention to name heavy tests like this *LiveTest, which will exclude running it from the default CI build:
```suggestion
class CustomerResourceLiveTest {
```

File: quarkus-modules/quarkus-testcontainers/customer-service/src/test/java/info/customer/CustomerResourceTest.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    @ParameterizedTest
+    @MethodSource(value = "customerDataProvider")
+    void findById(long customerId, String customerName, int orderSize) {
+        RestAssured.given().pathParam("id", customerId)
+                .get()
+                .then().statusCode(200);
```
Comment (theangrydev) [inline quarkus-modules/quarkus-testcontainers/customer-service/src/test/java/info/customer/CustomerResourceTest.java:24]: We have a standard formatter to use for Java code, can you apply it throughout please:
https://github.com/Baeldung/dev-settings/tree/main/javaCodeStyle/intellij

**Review by theangrydev — CHANGES_REQUESTED** [review]
I'm assuming most of this is default Quarkus boilerplate. Can we trim out any unused boilerplate maybe to make it more clear which key files are actually relevant? e.g. just one Dockerfile, remove index.html, ...

**Comment by theangrydev** [comment]
Seeing this in CI:
```
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:3.8.2:build (default) on project order-service: Execution default of goal io.quarkus.platform:quarkus-maven-plugin:3.8.2:build failed: Unable to load the mojo 'build' (or one of its required components) from the plugin 'io.quarkus.platform:quarkus-maven-plugin:3.8.2': com.google.inject.ProvisionException: Unable to provision, see the following errors:
[ERROR] 
[ERROR] 1) No implementation for io.quarkus.maven.QuarkusBootstrapProvider was bound.
[ERROR] while locating io.quarkus.maven.BuildMojo
[ERROR] at ClassRealm[extension>io.quarkus.platform:quarkus-maven-plugin:3.8.2, parent: jdk.internal.loader.ClassLoaders$AppClassLoader@30946e09] (via modules: org.eclipse.sisu.wire.WireModule -> org.eclipse.sisu.plexus.PlexusBindingModule)
[ERROR] while locating org.apache.maven.plugin.Mojo annotated with @com.google.inject.name.Named(value="io.quarkus.platform:quarkus-maven-plugin:3.8.2:build")
[ERROR] 
[ERROR] 1 error
[ERROR] role: org.apache.maven.plugin.Mojo
[ERROR] roleHint: io.quarkus.platform:quarkus-maven-plugin:3.8.2:build
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginContainerException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :order-service
```

**Comment by pauldpearson** [comment]
@theangrydev I'm not sure whats going on here but I am fine locally and was able to resolve the issue in the spring-project module.  Not sure if its the Java version Jenkins is using that is causing issues or not.  It looks like it is using 17.0.1 and I'd think the quarkus-maven-plugin 3.8.2 wouldn't take issue with that.

**Comment by theangrydev** [comment]
@pauldpearson ok let me ping someone on our side to investigate, feel free to proceed with the article in the meantime if you're ready to

**Comment by theangrydev** [comment]
https://github.com/quarkusio/quarkus/issues/32379 might be relevant

**Comment by theangrydev** [comment]
@pauldpearson we use Maven 3.6.0 in CI and that seems to be the cause of the error - I'm asking the team if we can upgrade to 3.9.6

**Comment by pauldpearson** [comment]
@theangrydev Okay sounds good.  If not then we can try downgrading Quarkus?

**Comment by asif-anwar** [comment]
retest this please

**Comment by theangrydev** [comment]
@asif-anwar one thing I'm still missing - what changes do we need to make here to respect the usual LiveTest skip mechanism?

**Comment by theangrydev** [comment]
Never mind, I can see the parent is not being specified in the two child modules, that should do it

File: quarkus-modules/quarkus-testcontainers/order-service/pom.xml [hunk 4]
```diff
@@ -0,0 +1,143 @@
+<?xml version="1.0"?>
+<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+    <modelVersion>4.0.0</modelVersion>
+    <groupId>info.order</groupId>
+    <artifactId>order-service</artifactId>
+    <version>1.0.0-SNAPSHOT</version>
```
Comment (theangrydev) [inline quarkus-modules/quarkus-testcontainers/order-service/pom.xml:7]: We need to add the parent module for our CI build to work properly:
```suggestion
    <modelVersion>4.0.0</modelVersion>
    <groupId>info.order</groupId>
    <artifactId>order-service</artifactId>
    <version>1.0.0-SNAPSHOT</version>

  <parent>
    <groupId>com.baeldung</groupId>
    <artifactId>quarkus-testcontainers</artifactId>
    <version>1.0.0-SNAPSHOT</version>
  </parent>
```

File: quarkus-modules/quarkus-testcontainers/pom.xml [hunk 5]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0"?>
+<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd"
+         xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>quarkus-testcontainers</artifactId>
+    <version>1.0-SNAPSHOT</version>
```
Comment (theangrydev) [inline quarkus-modules/quarkus-testcontainers/pom.xml:7]:
```suggestion
    <version>1.0.0-SNAPSHOT</version>
```

File: quarkus-modules/quarkus-vs-springboot/spring-project/pom.xml [hunk 6]
```diff
@@ -92,6 +92,11 @@
                 <groupId>org.springframework.boot</groupId>
                 <artifactId>spring-boot-maven-plugin</artifactId>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <version>3.2.5</version>
+            </plugin>
```
Comment (theangrydev) [inline quarkus-modules/quarkus-vs-springboot/spring-project/pom.xml:99]:
```suggestion
```

File: quarkus-modules/quarkus-testcontainers/customer-service/pom.xml [hunk 7]
```diff
@@ -0,0 +1,164 @@
+<?xml version="1.0"?>
+<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd"
+         xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+    <modelVersion>4.0.0</modelVersion>
+    <groupId>info.customer</groupId>
+    <artifactId>customer-service</artifactId>
+    <version>1.0.0-SNAPSHOT</version>
```
Comment (theangrydev) [inline quarkus-modules/quarkus-testcontainers/customer-service/pom.xml:8]: We need to add the parent module for our CI build to work properly:
```suggestion
    <modelVersion>4.0.0</modelVersion>
    <groupId>info.customer</groupId>
    <artifactId>customer-service</artifactId>
    <version>1.0.0-SNAPSHOT</version>

    <parent>
        <groupId>com.baeldung</groupId>
        <artifactId>quarkus-testcontainers</artifactId>
        <version>1.0.0-SNAPSHOT</version>
    </parent>
```

**Review by theangrydev — CHANGES_REQUESTED** [review]

**Comment by pauldpearson** [comment]
🤞 Pushed up suggested changes

**Comment by theangrydev** [comment]
@pauldpearson quarkus-modules/quarkus-vs-springboot/spring-project/pom.xml is a module for another article, so I wouldn't expect any changes at all there vs master

**Comment by theangrydev** [comment]
@asif-anwar we are now seeing:
```
[INFO] quarkus-project 0.1-SNAPSHOT ....................... SUCCESS [ 15.408 s]
[INFO] spring-project 0.1-SNAPSHOT ........................ FAILURE [  3.196 s]
[INFO] quarkus-funqy 1.0.0-SNAPSHOT ....................... SKIPPED
[INFO] quarkus-testcontainers 1.0.0-SNAPSHOT .............. SKIPPED
[INFO] order-service 1.0.0-SNAPSHOT ....................... SKIPPED
[INFO] customer-service 1.0.0-SNAPSHOT .................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:43 min
[INFO] Finished at: 2024-03-27T12:25:27Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.2:test (default-test) on project spring-project: Execution default-test of goal org.apache.maven.plugins:maven-surefire-plugin:2.22.2:test failed: java.lang.NoClassDefFoundError: org/junit/platform/launcher/core/LauncherFactory: org.junit.platform.launcher.core.LauncherFactory -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :spring-project
```

Failure at module unrelated to this PR spring-project

This also reproduces on origin/master (without these PR changes) with mvn 3.6.3 and java 17.0.10

**Comment by theangrydev** [comment]
@pauldpearson I'm not clear how quarkus-modules/quarkus-vs-springboot/spring-project/pom.xml was working before but looks like your change does fix it
```
<plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-surefire-plugin</artifactId>
                <version>3.2.5</version>
            </plugin>
```

So let's go back to that sorry - and extract a `${maven-surefire-plugin.version}` property for the version there

**Review by theangrydev — APPROVED** [review]

**Comment by pauldpearson** [comment]
Victory!

**Comment by theangrydev** [comment]
retest this

**Review by theangrydev — APPROVED** [review]

**Comment by theangrydev** [comment]
retest this please

**Comment by pauldpearson** [comment]
@theangrydev I'm getting the same failure locally as what is showing in Jenkins

`[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.2.5:test (default-test) on project spring-project: Execution default-test of goal org.apache.maven.plugins:maven-surefire-plugin:3.2.5:test failed: org.junit.platform.commons.JUnitException: TestEngine with ID 'junit-jupiter' failed to discover tests: ClassSelector [className = 'com.baeldung.spring_project.SpringContextTest'] resolution failed: 'java.util.stream.Stream org.junit.platform.commons.support.ReflectionSupport.streamNestedClasses(java.lang.Class, java.util.function.Predicate)' -> [Help 1]`

**Comment by theangrydev** [comment]
retest this please

**Comment by theangrydev** [comment]
retest this please

**Review by theangrydev — APPROVED** [review]

**Pull request merged**